x-pack/plugins/ml/common/types/ml_url_generator.ts

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { RefreshInterval, TimeRange } from '../../../../../src/plugins/data/common/query';
+import { JobId } from '../../../reporting/common/types';
+import { ML_PAGES } from '../constants/ml_url_generator';
+
+export interface MLPageState<PageType, PageState> {
+  page: PageType;
+  pageState?: PageState;
```

Comment:
If I'm not mistaken for some pages `pageState` is required. Hence it would be better to utilize a [conditional type](https://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types) instead because we want to provide a better experience for the users of this URL generator 🙂 . 
It's going to look something like this, but probably could be improved 
```suggestion
export type MLPageState<PageType, PageState = undefined> = PageState extends undefined
  ? { page: PageType }
  : { page: PageType; pageState: PageState }
```
In case you want `pageState` to be optional, just provide init it adding `undefined` to the union, i.e.
```typescript
export type SomePageState = MLPageState<typeof ML_PAGES.SOME_PAGE, PageStateType | undefined>;

Reply:
Updated here https://github.com/elastic/kibana/pull/75696/commits/9962c25afba12c5d6ebc699f1134f010176b068f